8308614: Enabling JVMTI ClassLoad event slows down vthread creation by factor 10

sspitsyn opened the pull request:
This is a fix of a performance/scalability related issue. The `JvmtiThreadState` objects for virtual thread filtered events enabled globally are created eagerly because it is needed when the `interp_only_mode` is enabled. Otherwise, some events which are generated in `interp_only_mode` from the debugging version of interpreter chunks can be missed.
However, it has to be okay to avoid eager creation of these object if no `interp_only_mode` has ever been requested.
It seems to be an extremely important optimization to create JvmtiThreadState objects lazily in such cases.
It is done by introducing the flag `JvmtiThreadState::_seen_interp_only_mode` which indicates when the `JvmtiThreadState` objects have to be created eagerly.

Additionally, the fix includes the following related changes:
 - Use condition double checking idiom for `MutexLocker mu(JvmtiThreadState_lock)` in the function `JvmtiVTMSTransitionDisabler::VTMS_mount_end` which is on a performance-critical path and looks like this:
```
  JvmtiThreadState* state = thread->jvmti_thread_state();
  if (state != nullptr && state->is_pending_interp_only_mode()) {
    MutexLocker mu(JvmtiThreadState_lock);
    state = thread->jvmti_thread_state();
    if (state != nullptr && state->is_pending_interp_only_mode()) {
      JvmtiEventController::enter_interp_only_mode();
    }
  }
```

 - Add extra check of `JvmtiExport::can_support_virtual_threads()` when virtual thread mount and unmount are posted.
 - Minor: Added a `ThreadsListHandle` to the `JvmtiEventControllerPrivate::enter_interp_only_mode`. It is needed because of the dynamic creation of compensating carrier threads which is racy for JVMTI `SetEventNotificationMode` implementation.

Performance mesurements:
- Without this fix the test provided by the bug submitter gives execution numbers:
  - no ClassLoad events enabled:     3251 ms
  - ClassLoad events are enabled: 40534 ms
  
- With the fix:
  - no ClassLoad events enabled:    3270 ms
  - ClassLoad events are enabled:   3385 ms
  
Testing:
 - Ran mach5 tiers 1-6, no regressions are noticed

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8308614](https://bugs.openjdk.org/browse/JDK-8308614): Enabling JVMTI ClassLoad event slows down vthread creation by factor 10 (**Bug** - P3)


### Reviewers
 * [Daniel D. Daugherty](https://openjdk.org/census#dcubed) (@dcubed-ojdk - **Reviewer**)
 * [Chris Plummer](https://openjdk.org/census#cjplummer) (@plummercj - **Reviewer**)
 * [Alex Menkov](https://openjdk.org/census#amenkov) (@alexmenkov - **Reviewer**)


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/16686/head:pull/16686` \
`$ git checkout pull/16686`

Update a local copy of the PR: \
`$ git checkout pull/16686` \
`$ git pull https://git.openjdk.org/jdk.git pull/16686/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 16686`

View PR using the GUI difftool: \
`$ git pr show -t 16686`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/16686.diff">https://git.openjdk.org/jdk/pull/16686.diff</a>

</details>


### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/16686#issuecomment-1814250953)

plummercj commented on src/hotspot/share/prims/jvmtiThreadState.hpp:
```diff
@@ -231,6 +231,8 @@ class JvmtiThreadState : public CHeapObj<mtInternal> {
   inline JvmtiEnvThreadState* head_env_thread_state();
   inline void set_head_env_thread_state(JvmtiEnvThreadState* ets);
 
+  static bool _seen_interp_only_mode; // needed for optimization
```
Say what the flag represents, not why we have it.

sspitsyn replied:
Thank you for looking at this PR!
Okay, thanks. Will do.

sspitsyn replied:
Fixed.

plummercj commented on src/hotspot/share/prims/jvmtiThreadState.cpp:
```diff
@@ -569,7 +575,8 @@ JvmtiVTMSTransitionDisabler::VTMS_vthread_mount(jobject vthread, bool hide) {
   } else {
     VTMS_mount_end(vthread);
-    if (JvmtiExport::should_post_vthread_mount()) {
+    if (JvmtiExport::can_support_virtual_threads() &&
+        JvmtiExport::should_post_vthread_mount()) {
```
It seems odd that "can_support" can be false when "should_post" is true. I would think that "should_post" would always be false when "can_support" is false, and therefore there would be no need to check "can_support".

sspitsyn replied:
Right, thanks. It is why this check was missed in the first place. Will undo this change.

sspitsyn replied:
Fixed.

plummercj commented on src/hotspot/share/prims/jvmtiThreadState.cpp:
```diff
@@ -526,15 +527,20 @@ JvmtiVTMSTransitionDisabler::VTMS_vthread_start(jobject vthread) {
   assert(!thread->is_in_tmp_VTMS_transition(), "sanity check");
 
-  JvmtiEventController::thread_started(thread);
+  // JvmtiThreadState objects for virtual thread filtered events enabled globally
+  // must be created eagerly if the interp_only_mode is enabled. Otherwise,
```
This sentence is hard to read. How about:

"If interp_only_mode is enabled then we must eagerly create JvmtiThreadState objects for globally enabled virtual thread filtered events."

sspitsyn replied:
Okay, thanks. The suggestion is taken.

sspitsyn replied:
Fixed.

plummercj commented on src/hotspot/share/prims/jvmtiThreadState.hpp:
```diff
@@ -250,6 +252,13 @@ class JvmtiThreadState : public CHeapObj<mtInternal> {
 
   static void periodic_clean_up();
 
+  // JvmtiThreadState objects for virtual thread filtered events enabled globally
+  // must be created eagerly if the interp_only_mode is enabled. Otherwise,
+  // it is an important optimization to create JvmtiThreadState objects lazily.
```
No need for this comment here. It is already at the call site, which is where it belongs. Instead the comment here should say what this API does (return true if any thread has entered interp_only_mode at any point during the JVMs execution).

sspitsyn replied:
Thanks, good suggestion.

sspitsyn replied:
Fixed.

plummercj commented on src/hotspot/share/prims/jvmtiEventController.cpp:
```diff
@@ -369,6 +369,7 @@ void JvmtiEventControllerPrivate::enter_interp_only_mode(JvmtiThreadState *state
   if (target == nullptr) { // an unmounted virtual thread
     return;  // EnterInterpOnlyModeClosure will be executed right after mount.
   }
+  ThreadsListHandle tlh(current);
```
Why was this added?

sspitsyn replied:
This is explained in the PR description.
Do you think, a just comment is needed or this has to be separated from this fix?

plummercj replied:
I see the PR comment, but I don't really understand it. Is this to force some sort of early initialization to avoid a race later on? It just seems odd to create the tlh, but never use it.

sspitsyn replied:
The `tlh` is used to protect any existing at this point JavaThread's from being terminated while it is set.
My understanding is that there is no need to iterate over all threads in the list to get this protection.

plummercj replied:
Ok. I comment indicating that purpose would be useful.

sspitsyn replied:
Okay, thanks!
I've added a comment:
```
  // Protects any existing JavaThread's from being terminated while it is set.
  // The FJP carrier thread compensating mechanism can create carrier threads concurrently.
```

dcubed-ojdk commented on src/hotspot/share/prims/jvmtiEventController.cpp:
```diff
@@ -369,6 +369,9 @@ void JvmtiEventControllerPrivate::enter_interp_only_mode(JvmtiThreadState *state
   if (target == nullptr) { // an unmounted virtual thread
     return;  // EnterInterpOnlyModeClosure will be executed right after mount.
   }
+  // Protects any existing JavaThread's from being terminated while it is set.
+  // The FJP carrier thread compensating mechanism can create carrier threads concurrently.
+  ThreadsListHandle tlh(current);
```
A ThreadsListHandle does not prevent a JavaThread from being terminated. It
prevents a JavaThread from exiting and being freed. The JavaThread is able to
set the terminated state on itself, but will not be able to complete exiting while
it is on a ThreadsListHandle. There is a subtle difference.

There's a `target` JavaThread that is fetched from a `JvmtiThreadState` object
and that `target` JavaThread is only protected by this `tlh` if `target` is included
in the ThreadsList that was captured by this `tlh`. In all likelihood, there should be
a ThreadsListHandle farther up the stack that's protecting the JavaThread from
which the `JvmtiThreadState` object was extracted and passed to this function.

As for carrier threads, if they are created _after_ this `tlh` was created, then this
`tlh` cannot protect them because they won't be on this `tlh`'s ThreadsList.

sspitsyn replied:
Thank you for the comment, Dan!
Agreed, the comment needs to be corrected in two aspects.
I tried to simplify it but failed to do it correctly.
It is interesting that there is a `ThreadsListHandle` farther up the stack but it does not help sometimes.
It is observed that a `JavaThread` (of a carrier thread) referenced from the  `JvmtiThreadState` object can be just created, so we need a `ThreadsListHandle` to avoid possible asserts. With this fix in place I do not see the asserts fired anymore.

dcubed-ojdk replied:
@sspitsyn - Please point me at the code where these JavaThreads are newly created?

sspitsyn replied:
@dcubed-ojdk
I don't know FJP good enough to point at the code where it happens. However, I observe that new `JavaThread `is being created between two points of the execution path.
 - First point is in the `JvmtiEventControllerPrivate::recompute_enabled()` at the line where a `ThreadsListHandle` is set. I've added a trap checking if any `JavaThread` pointed by `state->get_thread()` is not protected by the `tlh`. I can see this trap is not fired (I can't say it has never been fired).
 - Second point is in the `JvmtiEventControllerPrivate::enter_interp_only_mode()`. If  a `ThreadsListHandle` is NOT set then I can observe a `JavaThread` referenced by the state->get_thread() which is not protected by any TLH. It a TLH added into `JvmtiEventControllerPrivate::enter_interp_only_mode()` then this `JavaThread` is observed as protected by TLH.
I can provide a full stack trace for this `JavaThread` consisting of two parts: carrier thread and virtual thread frames. The name of carrier thread is  `ForkJoinPool-1-worker-1`. The virtual thread is a tested virtual thread. 
The thread dump looks las below:
```
DBG: enter_interp_only_mode: target: 0x7f93f8043d00  <NO-NAME> virt: 1 carrier: ForkJoinPool-1-worker-1
DBG: ##### NATIVE stacktrace of JavaThread: 0x7f93f8043d00 #####
Native frames: (J=compiled Java code, j=interpreted, Vv=VM code, C=native code)
j  fieldacc02.check(Ljava/lang/Object;)I+0
j  fieldacc02.lambda$testVirtualThread$0()V+12
j  fieldacc02$$Lambda+0x00007f943b001428.run()V+0
j  java.lang.Thread.runWith(Ljava/lang/Object;Ljava/lang/Runnable;)V+5 java.base@22-internal
j  java.lang.VirtualThread.run(Ljava/lang/Runnable;)V+66 java.base@22-internal
j  java.lang.VirtualThread$VThreadContinuation$1.run()V+8 java.base@22-internal
j  jdk.internal.vm.Continuation.enter0()V+4 java.base@22-internal
j  jdk.internal.vm.Continuation.enter(Ljdk/internal/vm/Continuation;Z)V+1 java.base@22-internal
J 124  jdk.internal.vm.Continuation.enterSpecial(Ljdk/internal/vm/Continuation;ZZ)V java.base@22-internal (0 bytes) @ 0x00007f94c7cdf744 [0x00007f94c7cdf5e0+0x0000000000000164]
j  jdk.internal.vm.Continuation.run()V+122 java.base@22-internal
j  java.lang.VirtualThread.runContinuation()V+70 java.base@22-internal
j  java.lang.VirtualThread$$Lambda+0x00007f943b0496c0.run()V+4 java.base@22-internal
j  java.util.concurrent.ForkJoinTask$RunnableExecuteAction.compute()Ljava/lang/Void;+4 java.base@22-internal
j  java.util.concurrent.ForkJoinTask$RunnableExecuteAction.compute()Ljava/lang/Object;+1 java.base@22-internal
j  java.util.concurrent.ForkJoinTask$InterruptibleTask.exec()Z+51 java.base@22-internal
j  java.util.concurrent.ForkJoinTask.doExec()V+10 java.base@22-internal
j  java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(Ljava/util/concurrent/ForkJoinTask;Ljava/util/concurrent/ForkJoinPool$WorkQueue;I)V+49 java.base@22-internal
j  java.util.concurrent.ForkJoinPool.scan(Ljava/util/concurrent/ForkJoinPool$WorkQueue;JI)J+271 java.base@22-internal
j  java.util.concurrent.ForkJoinPool.runWorker(Ljava/util/concurrent/ForkJoinPool$WorkQueue;)V+68 java.base@22-internal
j  java.util.concurrent.ForkJoinWorkerThread.run()V+31 java.base@22-internal
v  ~StubRoutines::call_stub 0x00007f94c7505d21
V  [libjvm.so+0xe7b719]  JavaCalls::call_helper(JavaValue*, methodHandle const&, JavaCallArguments*, JavaThread*)+0x4a9  (javaCalls.cpp:415)
V  [libjvm.so+0xe7bdd5]  JavaCalls::call_virtual(JavaValue*, Klass*, Symbol*, Symbol*, JavaCallArguments*, JavaThread*)+0x345  (javaCalls.cpp:329)
V  [libjvm.so+0xe7bff6]  JavaCalls::call_virtual(JavaValue*, Handle, Klass*, Symbol*, Symbol*, JavaThread*)+0x76  (javaCalls.cpp:191)
V  [libjvm.so+0xfd9723]  thread_entry(JavaThread*, JavaThread*)+0x93  (jvm.cpp:2937)
V  [libjvm.so+0xeb06ac]  JavaThread::thread_main_inner()+0xcc  (javaThread.cpp:720)
V  [libjvm.so+0x1789496]  Thread::call_run()+0xb6  (thread.cpp:220)
V  [libjvm.so+0x1493b27]  thread_native_entry(Thread*)+0x127  (os_linux.cpp:787)
```

The observed `JavaThead` does not look as a garbage because no crashes has ever been observed. Apparently, it has been recently created because it is not protected by the TLH which was set in the `JvmtiEventControllerPrivate::recompute_enabled()`. I guess, it has to be possible to find out where exactly in the FJP code it is created but it will take time. I'm not sure why do you need it though.

sspitsyn replied:
The stack trace of current thread (where the assert was fired) can explain what is going on a little bit:
```
Current thread (0x00007f93f8043d00):  JavaThread "ForkJoinPool-1-worker-1" daemon [_thread_in_vm, id=16779, stack(0x00007f948a597000,0x00007f948a697000) (1024K)]

Stack: [0x00007f948a597000,0x00007f948a697000],  sp=0x00007f948a6949e0,  free space=1014k
Native frames: (J=compiled Java code, j=interpreted, Vv=VM code, C=native code)
V  [libjvm.so+0x117937d]  JvmtiEventControllerPrivate::enter_interp_only_mode(JvmtiThreadState*)+0x45d  (jvmtiEventController.cpp:402)
V  [libjvm.so+0x1179520]  JvmtiEventControllerPrivate::recompute_thread_enabled(JvmtiThreadState*) [clone .part.0]+0x190  (jvmtiEventController.cpp:632)
V  [libjvm.so+0x117a1e1]  JvmtiEventControllerPrivate::thread_started(JavaThread*)+0x351  (jvmtiEventController.cpp:1174)
V  [libjvm.so+0x117e608]  JvmtiExport::get_jvmti_thread_state(JavaThread*)+0x98  (jvmtiExport.cpp:424)
V  [libjvm.so+0x118a86c]  JvmtiExport::post_field_access(JavaThread*, Method*, unsigned char*, Klass*, Handle, _jfieldID*)+0x6c  (jvmtiExport.cpp:2214)
V  [libjvm.so+0x118b3a1]  JvmtiExport::post_field_access_by_jni(JavaThread*, oop, Klass*, _jfieldID*, bool)+0x321  (jvmtiExport.cpp:2202)
V  [libjvm.so+0x118b4e9]  JvmtiExport::jni_GetField_probe(JavaThread*, _jobject*, oop, Klass*, _jfieldID*, bool)+0x79  (jvmtiExport.cpp:2168)
V  [libjvm.so+0xf83847]  jni_GetStaticBooleanField+0x257  (jni.cpp:2047)
C  [libfieldacc02.so+0x379b]  Java_fieldacc02_check+0x6b  (jni.h:1546)
j  fieldacc02.check(Ljava/lang/Object;)I+0
j  fieldacc02.lambda$testVirtualThread$0()V+12
j  fieldacc02$$Lambda+0x00007f943b001428.run()V+0
j  java.lang.Thread.runWith(Ljava/lang/Object;Ljava/lang/Runnable;)V+5 java.base@22-internal
j  java.lang.VirtualThread.run(Ljava/lang/Runnable;)V+66 java.base@22-internal
j  java.lang.VirtualThread$VThreadContinuation$1.run()V+8 java.base@22-internal
j  jdk.internal.vm.Continuation.enter0()V+4 java.base@22-internal
j  jdk.internal.vm.Continuation.enter(Ljdk/internal/vm/Continuation;Z)V+1 java.base@22-internal
J 124  jdk.internal.vm.Continuation.enterSpecial(Ljdk/internal/vm/Continuation;ZZ)V java.base@22-internal (0 bytes) @ 0x00007f94c7cdf744 [0x00007f94c7cdf5e0+0x0000000000000164]
j  jdk.internal.vm.Continuation.run()V+122 java.base@22-internal
j  java.lang.VirtualThread.runContinuation()V+70 java.base@22-internal
j  java.lang.VirtualThread$$Lambda+0x00007f943b0496c0.run()V+4 java.base@22-internal
j  java.util.concurrent.ForkJoinTask$RunnableExecuteAction.compute()Ljava/lang/Void;+4 java.base@22-internal
j  java.util.concurrent.ForkJoinTask$RunnableExecuteAction.compute()Ljava/lang/Object;+1 java.base@22-internal
j  java.util.concurrent.ForkJoinTask$InterruptibleTask.exec()Z+51 java.base@22-internal
j  java.util.concurrent.ForkJoinTask.doExec()V+10 java.base@22-internal
j  java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(Ljava/util/concurrent/ForkJoinTask;Ljava/util/concurrent/ForkJoinPool$WorkQueue;I)V+49 java.base@22-internal
j  java.util.concurrent.ForkJoinPool.scan(Ljava/util/concurrent/ForkJoinPool$WorkQueue;JI)J+271 java.base@22-internal
j  java.util.concurrent.ForkJoinPool.runWorker(Ljava/util/concurrent/ForkJoinPool$WorkQueue;)V+68 java.base@22-internal
j  java.util.concurrent.ForkJoinWorkerThread.run()V+31 java.base@22-internal
v  ~StubRoutines::call_stub 0x00007f94c7505d21
V  [libjvm.so+0xe7b719]  JavaCalls::call_helper(JavaValue*, methodHandle const&, JavaCallArguments*, JavaThread*)+0x4a9  (javaCalls.cpp:415)
V  [libjvm.so+0xe7bdd5]  JavaCalls::call_virtual(JavaValue*, Klass*, Symbol*, Symbol*, JavaCallArguments*, JavaThread*)+0x345  (javaCalls.cpp:329)
V  [libjvm.so+0xe7bff6]  JavaCalls::call_virtual(JavaValue*, Handle, Klass*, Symbol*, Symbol*, JavaThread*)+0x76  (javaCalls.cpp:191)
V  [libjvm.so+0xfd9723]  thread_entry(JavaThread*, JavaThread*)+0x93  (jvm.cpp:2937)
V  [libjvm.so+0xeb06ac]  JavaThread::thread_main_inner()+0xcc  (javaThread.cpp:720)
V  [libjvm.so+0x1789496]  Thread::call_run()+0xb6  (thread.cpp:220)
V  [libjvm.so+0x1493b27]  thread_native_entry(Thread*)+0x127  (os_linux.cpp:787)
```

dholmes-ora replied:
Just to re-iterate what Dan was saying, the TLH is only of use if you are accessing threads known to be included in the TLH.

sspitsyn replied:
The issue occurred to be with the current thread which has to be safe to access without a TLH.
However, there is this guarantee in the `Hanshake::execute()`:
```
void Handshake::execute(HandshakeClosure* hs_cl, ThreadsListHandle* tlh, JavaThread* target) {
  . . .
  guarantee(target != nullptr, "must be");
  if (tlh == nullptr) {
    guarantee(Thread::is_JavaThread_protected_by_TLH(target),
              "missing ThreadsListHandle in calling context.");
```
This guarantee fired for current thread is a source of confusion and mistakes.
Would it be a right thing to correct it?

sspitsyn replied:
I've spend some time trying to reproduce the original issue to provide some details to Dan but there is no luck yet. I'm sure the issue still exists but it is better to investigate and address it separately from this PR.
So, my plan is to remove this line with TLH from the `jvmtiEventController.cpp`. BTW, if it is still needed Alan promised to point to the code where the FJP compensating mechanism adds new carrier threads.

dcubed-ojdk commented:
I'm going to resurrect the failing guarantee() code and part of the stack trace that was removed
and yack a bit about this code path.

Here's the location of the failing guarantee():
```
void Handshake::execute(HandshakeClosure* hs_cl, ThreadsListHandle* tlh, JavaThread* target) {
  . . .
  guarantee(target != nullptr, "must be");
  if (tlh == nullptr) {
    guarantee(Thread::is_JavaThread_protected_by_TLH(target),
              "missing ThreadsListHandle in calling context.");
```

and here's part of the stack trace that got us here:
```
V  [libjvm.so+0x117937d]  JvmtiEventControllerPrivate::enter_interp_only_mode(JvmtiThreadState*)+0x45d  (jvmtiEventController.cpp:402)
V  [libjvm.so+0x1179520]  JvmtiEventControllerPrivate::recompute_thread_enabled(JvmtiThreadState*) [clone .part.0]+0x190  (jvmtiEventController.cpp:632)
V  [libjvm.so+0x117a1e1]  JvmtiEventControllerPrivate::thread_started(JavaThread*)+0x351  (jvmtiEventController.cpp:1174)
V  [libjvm.so+0x117e608]  JvmtiExport::get_jvmti_thread_state(JavaThread*)+0x98  (jvmtiExport.cpp:424)
V  [libjvm.so+0x118a86c]  JvmtiExport::post_field_access(JavaThread*, Method*, unsigned char*, Klass*, Handle, _jfieldID*)+0x6c  (jvmtiExport.cpp:2214)
```

This must have been a stack trace from a build with some optimizations enabled because
when I look at last night's code base, I see 8 frames from the JvmtiExport::get_jvmti_thread_state()
call to Handshake::execute() with three params:

```
src/hotspot/share/prims/jvmtiExport.cpp:
    JvmtiExport::get_jvmti_thread_state(JavaThread *thread) {
      assert(thread == JavaThread::current(), "must be current thread");
      if (thread->is_vthread_mounted() && thread->jvmti_thread_state() == nullptr) {
        JvmtiEventController::thread_started(thread);
      }
```
The above code asserts that the `thread` parameter is the current thread so
we know that the calling thread is operating on itself.

```
src/hotspot/share/prims/jvmtiEventController.cpp
    JvmtiEventControllerPrivate::thread_started(JavaThread *thread) {
      assert(thread == Thread::current(), "must be current thread");
<snip>
  // if we have any thread filtered events globally enabled, create/update the thread state
  if (is_any_thread_filtered_event_enabled_globally()) { // intentionally racy
    JvmtiThreadState::state_for(thread);
```
The above code asserts that the `thread` parameter is the current thread so
we know that the calling thread is operating on itself. Note that we're calling
the single parameter version of `JvmtiThreadState::state_for()` here and in
that case the `thread_handle` parameter is `Handle thread_handle = Handle()`.

```
src/hotspot/share/prims/jvmtiThreadState.inline.hpp
    inline JvmtiThreadState* JvmtiThreadState::state_for(JavaThread *thread, Handle thread_handle) {
      // In a case of unmounted virtual thread the thread can be null.
      JvmtiThreadState* state = thread_handle == nullptr ? thread->jvmti_thread_state() :
                                                    java_lang_Thread::jvmti_thread_state(thread_handle());
      if (state == nullptr) {
        MutexLocker mu(JvmtiThreadState_lock);
        // check again with the lock held
        state = state_for_while_locked(thread, thread_handle());
        JvmtiEventController::recompute_thread_filtered(state);
```
The above code grabs the JvmtiThreadState from the `thread` parameter and
passes that to the `JvmtiEventController::recompute_thread_filtered()` call.
We know that `thread` parameter is the current thread.

```
src/hotspot/share/prims/jvmtiEventController.cpp
    void
    JvmtiEventControllerPrivate::recompute_thread_filtered(JvmtiThreadState *state) {
<snip>
      if (is_any_thread_filtered_event_enabled_globally()) {
        JvmtiEventControllerPrivate::recompute_thread_enabled(state);
```
The above code is just a filter wrapper for calling `JvmtiEventControllerPrivate::recompute_thread_enabled(`.

```
src/hotspot/share/prims/jvmtiEventController.cpp
    jlong
    JvmtiEventControllerPrivate::recompute_thread_enabled(JvmtiThreadState *state) {
<snip>
      // compute interp_only mode
      bool should_be_interp = (any_env_enabled & INTERP_EVENT_BITS) != 0 || has_frame_pops;
      bool is_now_interp = state->is_interp_only_mode();

      if (should_be_interp != is_now_interp) {
        if (should_be_interp) {
          enter_interp_only_mode(state);
```
The above code determines that the current thread needs to be in
interpreted mode so it calls `enter_interp_only_mode(state)`.

```
src/hotspot/share/prims/jvmtiEventController.cpp
    void JvmtiEventControllerPrivate::enter_interp_only_mode(JvmtiThreadState *state) {
<snip>
      JavaThread *target = state->get_thread();
      Thread *current = Thread::current();
<snip>
      if (target->is_handshake_safe_for(current)) {
        hs.do_thread(target);
      } else {
        assert(state->get_thread() != nullptr, "sanity check");
        Handshake::execute(&hs, target);
```
When know from our previous code analysis that the `JvmtiThreadState *state`
we were passed was fetched from the current thread. See `JvmtiThreadState::state_for`
above. So that `target` thread and the `current` should be the same thread.

Why does this check return false:

    `if (target->is_handshake_safe_for(current)) {`

which allows us to travel down this call: `Handshake::execute(&hs, target)`

```
src/hotspot/share/runtime/handshake.cpp
    void Handshake::execute(HandshakeClosure* hs_cl, JavaThread* target) {
      // tlh == nullptr means we rely on a ThreadsListHandle somewhere
      // in the caller's context (and we sanity check for that).
      Handshake::execute(hs_cl, nullptr, target);
    }
```
The two parameter version of `Handshake::execute()` is just a wrapper
that passed a nullptr for the ThreadsListHandle to the three parameter
version of `Handshake::execute()`. And that's how we get to the failing
guarantee()...

```
src/hotspot/share/runtime/handshake.cpp
    void Handshake::execute(HandshakeClosure* hs_cl, ThreadsListHandle* tlh, JavaThread* target) {
      JavaThread* self = JavaThread::current();
      HandshakeOperation op(hs_cl, target, Thread::current());

      jlong start_time_ns = os::javaTimeNanos();

      guarantee(target != nullptr, "must be");
      if (tlh == nullptr) {
        guarantee(Thread::is_JavaThread_protected_by_TLH(target),
                  "missing ThreadsListHandle in calling context.");
        target->handshake_state()->add_operation(&op);
```

One possible fix for the guarantee is this version:
```
        guarantee(self == target || Thread::is_JavaThread_protected_by_TLH(target),
                  "missing ThreadsListHandle in calling context.");
```

However, that ignores why this code in JvmtiEventControllerPrivate::enter_interp_only_mode
return false:

    `if (target->is_handshake_safe_for(current)) {`

when we have these local variable values:
```
      JavaThread *target = state->get_thread();
      Thread *current = Thread::current();
```

```
src/hotspot/share/runtime/javaThread.hpp
      // A JavaThread can always safely operate on it self and other threads
      // can do it safely if they are the active handshaker.
      bool is_handshake_safe_for(Thread* th) const {
        return _handshake.active_handshaker() == th || this == th;
      }
```
It seems to me that this portion of the logic: `this == th` should have
returned `true` and not `false`. What am I missing here?

sspitsyn commented:
@dcubed-ojdk
Thank you for the analysis. I agree with it. It is why I've removed this stack trace and posted my statement:
> I've removed a part of this comment with stack traces as my traps were not fully correct, need to double check everything. > This issue is not well reproducible but I'm still trying to reproduce it again.

Initially, I added a trap to catch the issue with the JVMTI SetEventNotificationMode related code path and mistakenly thought that I've caught it. But it was a different code path that you just described. So. I've removed this part of my comment as it was wrong. Then I tried to reproduce the code path I wanted to catch but had no luck to reproduce it.
Now, I'm going to remove this line of code from my PR we are discussing.
I'm suggesting to investigate the issue with guarantee separately from this PR.
Is it okay with you?

dcubed-ojdk commented:
Removing the TLH is the right thing to do. If we get that failure mode again, then we
can file a new bug and hopefully have an hs_err_pid file with it.

I don't think we should change the guarantee() in `Handshake::execute()`. When the
three parameter version of `execute()` is called with `tlh == nullptr`, the caller is
saying that there is supposed to be a ThreadsListHandle in the calling context. Yes,
if the target thread is the calling thread, then a ThreadsListHandle is not needed,
but that's why we have this code to prevent the call to `Handshake::execute()`:
```
      if (target->is_handshake_safe_for(current)) {
        hs.do_thread(target);
```

In other words, I think `Handshake::execute()` is working the way it is supposed to
when `tlh == nullptr` is passed.

dcubed-ojdk commented on src/hotspot/share/prims/jvmtiThreadState.cpp:
```diff
@@ -526,11 +527,14 @@ JvmtiVTMSTransitionDisabler::VTMS_vthread_start(jobject vthread) {
+  if (JvmtiThreadState::seen_interp_only_mode()) {
+    JvmtiEventController::thread_started(thread);
+  }
+  if (JvmtiExport::should_post_vthread_start()) {
```
Why is this check: `if (JvmtiExport::can_support_virtual_threads()) {` removed?

sspitsyn replied:
It is because the `JvmtiExport::should_post_vthread_start()` returns `true` only if the `JvmtiExport::can_support_virtual_threads()` returns `true`.

The JVMTI `SetEventNotificationMode()` function checks for a required event capability by using the `JvmtiUtil::has_event_capability()` function that is defined in the jvmtiEnter.cpp generated with the `jvmtiEnter.xsl` script.

It looks like this:
```
// Check Event Capabilities
bool JvmtiUtil::has_event_capability(jvmtiEvent event_type, const jvmtiCapabilities* capabilities_ptr) {
  switch (event_type) {
    . . .
    case JVMTI_EVENT_VIRTUAL_THREAD_START:
      return capabilities_ptr->can_support_virtual_threads != 0;
    case JVMTI_EVENT_VIRTUAL_THREAD_END:
      return capabilities_ptr->can_support_virtual_threads != 0;
    . . .
```

dcubed-ojdk commented on src/hotspot/share/prims/jvmtiThreadState.cpp:
```diff
@@ -549,10 +553,8 @@ JvmtiVTMSTransitionDisabler::VTMS_vthread_end(jobject vthread) {
   if (JvmtiExport::should_post_vthread_unmount()) {
     JvmtiExport::post_vthread_unmount(vthread);
   }
-  if (JvmtiExport::can_support_virtual_threads()) {
```
Why is this check: if (JvmtiExport::can_support_virtual_threads()) { removed?

sspitsyn replied:
The same reason as above. Thank you for checking this.

dcubed-ojdk commented on src/hotspot/share/prims/jvmtiThreadState.hpp:
```diff
@@ -231,6 +231,8 @@ class JvmtiThreadState : public CHeapObj<mtInternal> {
   inline JvmtiEnvThreadState* head_env_thread_state();
   inline void set_head_env_thread_state(JvmtiEnvThreadState* ets);
 
+  static bool _seen_interp_only_mode; // interp_only_mode was requested once
```
perhaps: s/requested once/requested at least once/

sspitsyn replied:
Thank you - fixed now.

dcubed-ojdk commented on src/hotspot/share/prims/jvmtiThreadState.cpp:
```diff
@@ -526,11 +527,14 @@ JvmtiVTMSTransitionDisabler::VTMS_vthread_start(jobject vthread) {
-    if (JvmtiExport::should_post_vthread_start()) {
-      JvmtiExport::post_vthread_start(vthread);
-    }
+  // If interp_only_mode is enabled then we must eagerly create JvmtiThreadState
```
typo: s/is enabled/has been enabled/

sspitsyn commented:
> Removing the TLH is the right thing to do. If we get that failure mode again, then we
can file a new bug and hopefully have an hs_err_pid file with it.

Agreed.

dcubed-ojdk reviewed (approved):
Thanks for fixing the comments.

sspitsyn commented:
Chris and Alex, thank you for review!

sspitsyn commented:
/integrate
